{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1000 Genomes Exercises in SciDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SciDB is a scale-out DBMS designed for scientific use cases, with built-in capabilities for sophisticated mathematical calculations beyond the typical domain of SQL/NoSQL systems. This workbook demonstrates using SciDB's R interface for scalable query and analysis of 1000 Genomes phase 3 genotype data, including simple aggregations such as computing transition/transversion ratios, and more advanced calculations such as principal component analysis and estimating linkage disequilibrium.\n",
    "\n",
    "First, we connect to SciDB and obtain handles to the data arrays. For development/testing on a modest machine, we'll just look at chromosomes 21 and 22 across the 2,504 individuals; but SciDB can scale out on a cluster to handle arbitrarily large datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "% matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "from scidbpy import connect\n",
    "sdb = connect()\n",
    "SAMPLE = sdb.wrap_array('KG_SAMPLE')\n",
    "CHROMOSOME = sdb.wrap_array('KG_CHROMOSOME')\n",
    "VARIANT = sdb.merge(sdb.wrap_array('KG_VARIANT'), \n",
    "                    sdb.afl.filter(\"KG_CHROMOSOME\", \"chromosome='10' or chromosome='20' or chromosome='21'\"))\n",
    "VARIANT.eval()\n",
    "GENOTYPE = sdb.afl.project(\n",
    "    sdb.merge(\n",
    "        sdb.wrap_array('KG_GENOTYPE'),\n",
    "        sdb.afl.project(VARIANT, 'chromosome')\n",
    "        ),\n",
    "    'allele_1', 'allele_2', 'phase'\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `VARIANT` array for chomosomes 21 and 22 has been stored in SciDB's memory (but not the memory of our local R process) by the `scidbeval()` expression above. In contrast, the `GENOTYPE` array is merely a lazy expression, denoting but not materializing the subset of all the genotypes on chromosomes 21 and 22. Because `GENOTYPE` is our biggest array by far, we avoid copying any significant portion of it.\n",
    "\n",
    "Let's take a look at the schema."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>alternate</th>\n",
       "      <th>id</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>info</th>\n",
       "      <th>ac</th>\n",
       "      <th>af</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromosome_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>alternate_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">9</th>\n",
       "      <th>60494</th>\n",
       "      <th>60494</th>\n",
       "      <th>1</th>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td>           .</td>\n",
       "      <td> 100</td>\n",
       "      <td> PASS</td>\n",
       "      <td> AC=27;AF=0.00539137;AN=5008;NS=2504</td>\n",
       "      <td> 27</td>\n",
       "      <td> 0.005391</td>\n",
       "      <td> 10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60515</th>\n",
       "      <th>60515</th>\n",
       "      <th>1</th>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td>           .</td>\n",
       "      <td> 100</td>\n",
       "      <td> PASS</td>\n",
       "      <td> AC=1;AF=0.000199681;AN=5008;NS=2504</td>\n",
       "      <td>  1</td>\n",
       "      <td> 0.000200</td>\n",
       "      <td> 10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60523</th>\n",
       "      <th>60523</th>\n",
       "      <th>1</th>\n",
       "      <td> T</td>\n",
       "      <td> G</td>\n",
       "      <td> rs148087467</td>\n",
       "      <td> 100</td>\n",
       "      <td> PASS</td>\n",
       "      <td>  AC=91;AF=0.0181709;AN=5008;NS=2504</td>\n",
       "      <td> 91</td>\n",
       "      <td> 0.018171</td>\n",
       "      <td> 10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60607</th>\n",
       "      <th>60607</th>\n",
       "      <th>1</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>           .</td>\n",
       "      <td> 100</td>\n",
       "      <td> PASS</td>\n",
       "      <td> AC=1;AF=0.000199681;AN=5008;NS=2504</td>\n",
       "      <td>  1</td>\n",
       "      <td> 0.000200</td>\n",
       "      <td> 10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60684</th>\n",
       "      <th>60684</th>\n",
       "      <th>1</th>\n",
       "      <td> A</td>\n",
       "      <td> C</td>\n",
       "      <td>           .</td>\n",
       "      <td> 100</td>\n",
       "      <td> PASS</td>\n",
       "      <td>  AC=62;AF=0.0123802;AN=5008;NS=2504</td>\n",
       "      <td> 62</td>\n",
       "      <td> 0.012380</td>\n",
       "      <td> 10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       reference alternate           id  qual  \\\n",
       "chromosome_id start end   alternate_id                                          \n",
       "9             60494 60494 1                    A         G            .   100   \n",
       "              60515 60515 1                    C         T            .   100   \n",
       "              60523 60523 1                    T         G  rs148087467   100   \n",
       "              60607 60607 1                    G         A            .   100   \n",
       "              60684 60684 1                    A         C            .   100   \n",
       "\n",
       "                                       filter  \\\n",
       "chromosome_id start end   alternate_id          \n",
       "9             60494 60494 1              PASS   \n",
       "              60515 60515 1              PASS   \n",
       "              60523 60523 1              PASS   \n",
       "              60607 60607 1              PASS   \n",
       "              60684 60684 1              PASS   \n",
       "\n",
       "                                                                       info  \\\n",
       "chromosome_id start end   alternate_id                                        \n",
       "9             60494 60494 1             AC=27;AF=0.00539137;AN=5008;NS=2504   \n",
       "              60515 60515 1             AC=1;AF=0.000199681;AN=5008;NS=2504   \n",
       "              60523 60523 1              AC=91;AF=0.0181709;AN=5008;NS=2504   \n",
       "              60607 60607 1             AC=1;AF=0.000199681;AN=5008;NS=2504   \n",
       "              60684 60684 1              AC=62;AF=0.0123802;AN=5008;NS=2504   \n",
       "\n",
       "                                        ac        af chromosome  \n",
       "chromosome_id start end   alternate_id                           \n",
       "9             60494 60494 1             27  0.005391         10  \n",
       "              60515 60515 1              1  0.000200         10  \n",
       "              60523 60523 1             91  0.018171         10  \n",
       "              60607 60607 1              1  0.000200         10  \n",
       "              60684 60684 1             62  0.012380         10  \n",
       "\n",
       "[5 rows x 9 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VARIANT.todataframe().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'<reference:string,alternate:string,id:string NULL DEFAULT null,qual:double NULL DEFAULT null,filter:string NULL DEFAULT null,info:string NULL DEFAULT null,ac:double NULL DEFAULT null,af:double NULL DEFAULT null,chromosome:string> [chromosome_id=0:*,1,0,start=0:*,10000000,0,end=0:*,10000000,0,alternate_id=0:19,20,0]'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VARIANT.schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19864.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VARIANT.count().toarray()[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "      <th>phase</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           allele_1  allele_2  phase\n",
       "sample_id                           \n",
       "0                 0         0      1\n",
       "1                 0         0      1\n",
       "2                 0         0      1\n",
       "3                 0         0      1\n",
       "4                 0         0      1\n",
       "\n",
       "[5 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GENOTYPE[9,60494,60494,1,0:5].todataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'<allele_1:bool NULL DEFAULT null,allele_2:bool NULL DEFAULT null,phase:bool NULL DEFAULT null> [chromosome_id=0:*,1,0,start=0:*,10000000,0,end=0:*,10000000,0,alternate_id=0:19,20,0,sample_id=0:*,100,0]'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GENOTYPE.schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49739456"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(GENOTYPE.op_count().toarray()[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On the other hand, small arrays can easily be moved back and forth between SciDB and R, with one pitfall: SciDB arrays are zero-based, while R uses one-based indexing. But as genome scientists, we're right at home dealing with that!\n",
    "\n",
    "## Transition/transversion ratio\n",
    "\n",
    "The transition/transversion ratio (Ti/Tv) is a common quality metric for variant call sets. Let's compute Ti/Tv of all the variants with respect to the reference genome. This first calculation is on the variants only, not the individuals' genotypes, and thus involves only a modest amount of data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18840"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP = sdb.afl.filter(VARIANT, \"(reference='A' or reference='G' or reference='C' or reference='T') and \\\n",
    "                        (alternate='A' or alternate='G' or alternate='C' or alternate='T')\")\n",
    "snps = int(SNP.op_count().toarray()[0])\n",
    "snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13209"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annotate each SNP as to whether it's a transition (or else transversion)\n",
    "\n",
    "transitions_filter_str =   \"(reference='A' and alternate='G') or (reference='G' and alternate='A') or \\\n",
    "                            (reference='C' and alternate='T') or (reference='T' and alternate='C')\"\n",
    "SNP = sdb.afl.apply(SNP, \"is_transition\", \"bool(iif(%s, 1, 0))\" % transitions_filter_str)\n",
    "SNP = SNP.eval()\n",
    "\n",
    "# count transitions\n",
    "ti = int((sdb.afl.filter(SNP['is_transition'], \"is_transition = TRUE\").count()).toarray()[0])\n",
    "ti"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5631"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count transversions\n",
    "tv = int((sdb.afl.filter(SNP['is_transition'], \"is_transition = FALSE\").count()).toarray()[0])\n",
    "tv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.345764517847629"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# report Ti/Tv\n",
    "assert ti + tv == snps\n",
    "float(ti)/tv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's look at the distribution of Ti/Tv across the individuals in the population. The R notebook uses pipeR's `%>>%` operator to express a multi-step data processing pipeline, where `f(x) %>>% g(y)` denotes `g(f(x),y)`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAAEPCAYAAAC3NDh4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGnNJREFUeJzt3X3U3GV95/H3xyCrgBIQDY+WqFBIT7GARK26HVoX0a3A\n2hbRajlK/UOOT7t2S/DY5nbPbov0tLXrHj27Pm1qJRqrZVERCMrUhyqUZyREyO6ma3QTQcRn2KR8\n94/5Bcab/O577mSe7uT9OmdOrvnNNb/re09+M9+5rus3vytVhSRJu/K4SQcgSZpeJglJUiuThCSp\nlUlCktTKJCFJamWSkCS1GmmSSHJxkjuT3JHksiT/IsmhSdYnuTvJNUmWzqp/T5KNSc4YZWySpPll\nVL+TSHIs8EXgxKp6KMkngCuBXwLuq6pLk1wEHFJVq5KsAC4DTgOOAq4Fjq+qh0cSoCRpXqPsSfwQ\n2A4ckGQ/4ADgO8BZwJqmzhrgnKZ8NrC2qrZX1WZgE7ByhPFJkuYxsiRRVfcDfw78H3rJ4YGqWg8s\nq6ptTbVtwLKmfCSwpW8XW+j1KCRJEzKyJJHkmcDbgGPpJYCDkrymv071xrrmGu/ymiGSNEH7jXDf\nzwH+oaq+B5Dk08Dzga1JDq+qrUmOAL7b1P82cEzf849utv2cJCYOSdoNVZWFPmeUcxIbgecleWKS\nAC8GNgCfAc5v6pwPXN6UrwDOS7J/kuXAccANu9pxVU39bfXq1ROPwTiNc7HGaJzDv+2ukfUkquq2\nJH8N3Ag8DNwM/DfgScC6JBcAm4Fzm/obkqyjl0h2ABfWnvxlkqQ9NsrhJqrqUuDSWZvvp9er2FX9\nPwH+ZJQxSZIG5y+uR6TT6Uw6hIEY53AthjgXQ4xgnNNiZD+mG5UkjkJJ0gIloaZs4lqStMiZJCRJ\nrUwSkqRWJglJUiuThCSplUlCktTKJCFJamWSkCS1MklIklqZJCRJrUwSkqRWJglJUiuThCSplUlC\nktTKJCFJajXSlemk3dVbFn18XKNE2jWThKbYuD64x5uQpMVkpMNNSX4xyS19tx8keUuSQ5OsT3J3\nkmuSLO17zsVJ7kmyMckZo4xPkjS3sS1fmuRxwLeBlcCbgfuq6tIkFwGHVNWqJCuAy4DTgKOAa4Hj\nq+rhvv24fOk+oDfcNL6ehMeU9naLYfnSFwObqupbwFnAmmb7GuCcpnw2sLaqtlfVZmATvaQiSZqA\ncSaJ84C1TXlZVW1rytuAZU35SGBL33O20OtRSJImYCxJIsn+wMuBT85+rBk7mquv7ziAJE3IuM5u\neilwU1Xd29zfluTwqtqa5Ajgu832bwPH9D3v6Gbbz5mZmXmk3Ol06HQ6o4hZkhatbrdLt9vd4/2M\nZeI6yceBz1fVmub+pcD3qurdSVYBS2dNXK/k0YnrZ/XPVDtxvW9w4loart2duB55kkhyIPBPwPKq\n+lGz7VBgHfB0YDNwblU90Dz2DuD1wA7grVV19az9mST2ASYJabimNkkMm0li32CSkIZrMZwCK0la\nZEwSkqRWJglJUiuThCSplUlCktTKJCFJamWSkCS1MklIklqZJCRJrUwSkqRWJglJUiuThCSplUlC\nktTKJCFJamWSkCS1MklIklqZJCRJrUwSkqRWJglJUquRJ4kkS5P8bZK7kmxI8twkhyZZn+TuJNck\nWdpX/+Ik9yTZmOSMUccnSWo3jp7EXwFXVtWJwEnARmAVsL6qjge+0NwnyQrglcAK4EzgfUns7UjS\nhIz0AzjJwcCLqurDAFW1o6p+AJwFrGmqrQHOacpnA2urantVbQY2AStHGaMkqd2ov6UvB+5N8pEk\nNyf5QJIDgWVVta2psw1Y1pSPBLb0PX8LcNSIY5QktdhvDPs/BXhTVf1jkvfQDC3tVFWVpObYx2Me\nm5mZeaTc6XTodDpDCVaS9hbdbpdut7vH+0nVXJ/Pe7jz5HDga1W1vLn/QuBi4BnA6VW1NckRwHVV\ndUKSVQBVdUlT/ypgdVVd37fPGmXMmg5J2MX3g1G1hseU9nZJqKos9HkjHW6qqq3At5Ic32x6MXAn\n8Bng/Gbb+cDlTfkK4Lwk+ydZDhwH3DDKGCVJ7UY93ATwZuBjSfYH/ifwOmAJsC7JBcBm4FyAqtqQ\nZB2wAdgBXGi3QZImZ6TDTaPgcNO+weEmabimcrhJkrS4mSQkSa1MEpKkViYJSVIrk4QkqdU4ToHV\nXqB3tpGkfY1JQgswztNETUrSNHC4SZLUyiQhSWplkpAktTJJSJJamSQkSa1MEpKkViYJSVIrk4Qk\nqZVJQpLUyiQhSWplkpAktTJJSJJaLShJJDk0yUkLfM7mJLcnuSXJDX37WZ/k7iTXJFnaV//iJPck\n2ZjkjIW0JUkarnmTRJK/T/LkJIcCNwEfTPKXC2ijgE5VnVxVK5ttq4D1VXU88IXmPklWAK8EVgBn\nAu9LYm9HkiZkkA/gg6vqh8ArgL9uPuhfvMB2Zl/3+SxgTVNeA5zTlM8G1lbV9qraDGwCViJJmohB\nksSSJEcA5wKfa7YtZGGBAq5NcmOSNzTbllXVtqa8DVjWlI8EtvQ9dwtw1ALakiQN0SCLDv0H4Grg\nq1V1Q5JnAvcsoI0XVNX/TfJUYH2Sjf0PVlUlmSvpPOaxmZmZR8qdTodOp7OAcCRp79ftdul2u3u8\nn1SNb7WxJKuBHwNvoDdPsbXppVxXVSckWQVQVZc09a8CVlfV9X37qHHGrJ7e8qXjXpluXO0Fjynt\n7ZJQVQte8rE1SSR57xzPq6p6ywBBHQAsqaofJTkQuAZ4F705je9V1bubxLC0qlY1E9eX0ZuHOAq4\nFnhWf1YwSUyGSUJa3HY3Scw13HQTj75LZ+940HfUMuDveh8w7Ad8rKquSXIjsC7JBcBmevMdVNWG\nJOuADcAO4EIzgiRNzliHm4bBnsRk2JOQFrdR9CR27vhpwB/S++3CE5vNVVW/vtDGJEmLyyCnwH4M\n2Ag8A5ihNzx04+hCkiRNi3mHm5LcXFWnJLm9qk5qtt1YVc8ZS4SPjcfhpglwuEla3EY23AT8v+bf\nrUl+E/gOcMhCG5KmWXNyxViYkLSYDJIk/lNzAb63A+8Fngz825FGJY3d+Hot0mLi2U0ayN4+3OTQ\nlvZ2ozy76SOzNhVAVb1+oY1JkhaXQYabPsejX7OeCPwbevMSkqS93IKHm5r1Hb5aVc8fTUjztu9w\n0wQ43DS8tjx+NQm7O9y0Owv6HA88dTeeJ0laZAaZk/gxj37NKnrrP1w0yqAkSdNh3iRRVQeNIxBJ\n0vRpTRJJTmWOgdqqunkkEUmSpsZc60l06SWJJwKnArc3D50E3OjE9b7FievhteXxq0kY+sR1VXWq\n6nR6p7ueUlWnVtWpwMl4Cqwk7RMGObvphKq6Y+edqvoGcOLoQpIkTYtBfkx3e5IPAn9Dr1/+auC2\nkUYlSZoKg1wq/InAG4EXNZu+BLy/qh4ccWxt8TgnMQHOSQyvLY9fTcLuzkl4gT8NxCQxvLY8fjUJ\nQ5+4TvLJ5t9vJLlj1u32tuftYj9LktyS5DPN/UOTrE9yd5JrmsuQ76x7cZJ7kmxMcsZC/xhJ0nDN\ndQrskVX1nSS/wC4ugl9VmwdqIPl39E6hfVJVnZXkUuC+qro0yUXAIVW1KskK4DLgNOAo4Frg+Kp6\neNb+7ElMgD2J4bXl8atJGMUpsDtPc/1tYHtVbe6/DRjU0cDLgA/yaKI5C1jTlNcA5zTls4G1VbW9\n2f8mYOUC/hZJ0pANcgrsk4BrknwlyZuSLFvA/v8S+PdAf29gWVVta8rbgJ37OxLY0ldvC70ehSRp\nQga5dtMMMJPk2cC5wJeSbKmq35jrec162N+tqluSdFr2XUnm6nvv8rGZmZlHyp1Oh05nl7uXpH1W\nt9ul2+3u8X4GPrspyRH0hp5eBRxUVSfNU/9PgNcCO4An0Fsb+9P05hw6VbW12ed1VXVCklUAVXVJ\n8/yrgNVVdf2s/TonMQHOSQyvLY9fTcLI1pNIcmFzHacvAIcBvz9fggCoqndU1TFVtRw4D/hiVb0W\nuAI4v6l2PnB5U74COC/J/kmWA8cBNyz0D5IkDc8gv7h+OvC2qrp1D9va+fXpEmBdkguAzfSGsKiq\nDUnWARvo9T4utMsgSZM11ymwT66qHyZ5Crvoi1fV/aMOblccbpoMh5uG15bHryZhd4eb5upJrAX+\nNXATu34HLV9oY5KkxcXLcmgg9iSG15bHryZh6D2JJKfM9URXppOkvZ8r02kg9iSG15bHrybBlekk\nSUPnynSSpFauTCdJauXKdBqIcxLDa8vjV5PgynQaKZPE8Nry+NUkjOLHdDt3/EJgNXBsX/2qqmcs\ntDFJ0uIyyHDTN4G3ATcD/7xze1XdN9rQWuOxJzEB9iSG15bHryZhZD0J4IGq+vxuxCRJWuQG6Ulc\nAiyhtxbEQzu3T+oX1/YkJsOexPDa8vjVJIxs4rrvl9c/p/mh3diZJCbDJDG8tjx+NQme3aSRMkkM\nry2PX03CKC7w99qq+miSt/Pz76DQO7vpL3YjTknSIjLXxPUBzb9PYrxfISVJU8LhJg3E4abhteXx\nq0kY+lVgJUkaWZJI8oQk1ye5NcmGJH/abD80yfokdye5JsnSvudcnOSeJBuTnDGq2CRJgxnpcFOS\nA6rqp0n2A74C/AFwFnBfVV2a5CLgkKpalWQFcBlwGnAUcC1wfFU9PGufDjdNgMNNw2vL41eTMLLh\npiTv7Cs/YSE7r6qfNsX96f0g7/v0ksSaZvsa4JymfDawtqq2V9VmYBOwciHtSZKGqzVJJFmV5FeB\n3+nb/A8L2XmSxyW5FdgGXFdVdwLLqmpbU2UbsKwpHwls6Xv6Fno9CknShMx1CuxGeglieZKvAHcB\nhyU5oao2DrLzZqjoV5IcDFyd5PRZj1eSufreu3xsZmbmkXKn06HT6QwSjiTtM7rdLt1ud4/30zon\nkaQDfB34Gr15ghOBzwJfpLek6fMX1FDyR8DPgN8HOlW1NckR9HoYJyRZBVBVlzT1rwJWV9X1s/bj\nnMQEOCcxvLY8fjUJo5iTeAnwOeCZwJ/Tmx/4aVW9bpAEkeSwnWcuNavb/SvgFuAK4Pym2vnA5U35\nCuC8JPsnWQ4cB9yw0D9IkjQ8rcNNVXUxQJLbgI8Cp9IbbvoqcH9VvXyefR8BrEnyOHrJ6KNV9YUk\ntwDrklwAbAbObdrbkGQdsAHYAVxol0GSJmuQq8BeWlV/2JRvqaqTkzy1qu4dS4SPjcfcMQEONw2v\nLY9fTcJYrgKb5NlVddtCGxkmk8RkmCSG15bHrybBS4Xvg3of3ONkkhhGWx6/moRRLl+qqTbOD1JJ\n+xov8CdJamWSkCS1MklIklqZJCRJrZy4lsZsnGeleSaV9pRJQho7z0jT4uFwkySplUlCktTKJCFJ\namWSkCS1MklIklqZJCRJrUwSkqRWJglJUiuThCSplUlCktRqpEkiyTFJrktyZ5JvJHlLs/3QJOuT\n3J3kmiRL+55zcZJ7kmxMcsYo45MkzW2ky5cmORw4vKpuTXIQcBNwDvA64L6qujTJRcAhVbUqyQrg\nMuA04CjgWuD4qnq4b58uX9oY77rTrnG9GNvyvaKddnf50pH2JKpqa1Xd2pR/DNxF78P/LGBNU20N\nvcQBcDawtqq2V9VmYBOwcpQxSpLajW1OIsmxwMnA9cCyqtrWPLQNWNaUjwS29D1tC72kIkmagLFc\nKrwZavoU8Naq+lH/9fSrqpLM1Sd+zGMzMzOPlDudDp1OZ2ixStLeoNvt0u1293g/I52TAEjyeOCz\nwOer6j3Nto1Ap6q2JjkCuK6qTkiyCqCqLmnqXQWsrqrr+/bnnETDOQnbmq8t3yvaaSrnJNL7FPsQ\nsGFngmhcAZzflM8HLu/bfl6S/ZMsB44DbhhljJKkdqM+u+mFwJeA23n069PF9D741wFPBzYD51bV\nA81z3gG8HthBb3jq6ln7tCfRsCdhW/O15XtFO+1uT2Lkw03DZpJ4lEnCtuZry/eKdprK4SZJ0uJm\nkpAktTJJSJJamSQkSa1MEpKkViYJSVIrk4QkqZVJQpLUyiQhSWplkpAktTJJSJJamSQkSa1MEpKk\nViYJSVIrk4QkqZVJQpLUyiQhSWplkpAktdpv0gFIGp3eErfj43Kpe5+R9iSSfDjJtiR39G07NMn6\nJHcnuSbJ0r7HLk5yT5KNSc4YZWzSvqHGeNPeaNTDTR8Bzpy1bRWwvqqOB77Q3CfJCuCVwIrmOe9L\n4nCYJE3QSD+Eq+rLwPdnbT4LWNOU1wDnNOWzgbVVtb2qNgObgJWjjE+SNLdJfFNfVlXbmvI2YFlT\nPhLY0ldvC3DUOAOTJP28iU5cV1UlmWswc5ePzczMPFLudDp0Op3hBiZJi1y326Xb7e7xfjLqsxGS\nHAt8pqp+ubm/EehU1dYkRwDXVdUJSVYBVNUlTb2rgNVVdf2s/ZVnUPT0zlwZ12sxzrbG3Z5tDas9\n35vTKwlVteDT3SYx3HQFcH5TPh+4vG/7eUn2T7IcOA64YQLxSZIaIx1uSrIW+DXgsCTfAv4YuARY\nl+QCYDNwLkBVbUiyDtgA7AAutMsgSZM18uGmYXO46VEON9nW9LTVa8/35vRaTMNNkqRFwiQhSWpl\nkpAktTJJSJJamSQkSa1MEpKkViYJSVIrk4QkqZVJQpLUyiQhSWrlGteShmaca2p7CZDxMElIGqJx\nXpdK4+BwkySplUlCktTK4aYhuuOOO/jBD34w6TAkaWhcT2KITjvtdO66617222/pyNt66KEtPPjg\nP7E3r02wd/5te2tb427PtSsWanfXk7AnMUQ7dsBPfvJe4PQxtHYpcNEY2pG0L3NOQpLUauqSRJIz\nk2xMck8SvypL0gRNVZJIsgT4L8CZwArgVUlOnGxUu6s76QAG1J10AAPqTjqAAXUnHcAAupMOYEDd\nSQcwkG63O+kQRmqqkgSwEthUVZurajvwceDsCce0m7qTDmBA3UkHMKDupAMYUHfSAQygO+kABtSd\n89EkY7vNGaVJYqyOAr7Vd39Ls02SZqkx3fZt03Z206L+H1myBA488B0sWXIYDz74TZ7whJtG1tZD\nD23ioYdGtntJfebrTbzrXe8aanvTdHrvVP1OIsnzgJmqOrO5fzHwcFW9u6/O9AQsSYvI7vxOYtqS\nxH7AN4HfAL4D3AC8qqrummhgkrSPmqrhpqrakeRNwNXAEuBDJghJmpyp6klIkqbLtJ3dBECSDyfZ\nluSOlscPSfJ3SW5Lcn2SXxp3jE0cxyS5LsmdSb6R5C0t9f5z8+PA25KcPI1xJjkhydeSPJjk7eOO\ncQFx/m7zOt6e5KtJTprCGM9uYrwlyU1Jfn2cMQ4aZ1/d05LsSPKKccbYtD3I69lJ8oPm9bwlyTun\nMc6+WG9p6nTHHOagr+cf9L2WdzT/9+0XnKuqqbsBLwJOBu5oefzPgD9qyr8IXDuhOA8HfqUpH0Rv\nPuXEWXVeBlzZlJ8LfH1K43wq8BzgPwJvn+LX8/nAwU35zHG/ngPGeGBf+Zfp/fZn6l7L5rElwBeB\nzwK/NY1xAh3ginHHthtxLgXuBI5u7h82jXHOqv+b831+TmVPoqq+DHx/jionAtc1db8JHJvkqeOI\nrV9Vba2qW5vyj4G7gCNnVTsLWNPUuR5YmmTZtMVZVfdW1Y3A9nHGNiuGQeL8WlXtvB779cDRUxjj\nT/ruHgTcN74IH4lhkGMT4M3A3wL3jjG8RywgzokuRTdgnK8GPlVVW5p60/z/vtOrgbVz7XMqk8QA\nbgNeAZBkJfALjPnDYrYkx9Lr/Vw/66Fd/UBwYrHOEedUGTDOC4ArxxHPrswVY5JzktwFfB5oHeoZ\nh7Y4kxxF74oG7282TXSCco7Xs4BfbYbwrkyyYtyx9ZsjzuOAQ5vhnhuTvHbcsfWb7z2U5ADgJcCn\n5trPVJ3dtACXAH+V5BbgDuAW4J8nFUySg+h9G3trk70fU2XW/Ym8GQeIcyoMEmeS04HXAy8YZ2x9\n7c8ZY1VdDlye5EXAR+kNi47dPHG+B1hVVZXer8Um9m19njhvBo6pqp8meSlwOXD8uGOEeeN8PHAK\nvVP4DwC+luTrVXXPmMMc9L3+cuArVfXAXPtalEmiqn5E7wMCgCT/G/hfk4glyePpZeK/aT4YZvs2\ncEzf/aObbWM1QJxTYZA4m8nqDwBnVtVcw5IjsZDXsqq+nGS/JE+pqu+NJ8KeAeI8Ffh4Lz9wGPDS\nJNur6ooxhjlvnM37fWf580nel+TQqrp/muKkN2JwX1X9DPhZki8BzwbGmiQWcHyexzxDTbBIh5uS\nHJxk/6b8BuDvJ/HNuPn29SFgQ1W9p6XaFcDvNfWfBzxQVdvGFCJNu4PE+Uj1MYS064YHiDPJ04FP\nA6+pqk3jjK9pf5AYn9nUI8kpABNIEPPGWVXPqKrlVbWc3rfON04gQQzyei7rez1X0jt1f9wJYpD3\n0P8AXphkSTOU81xgw7hihMHf60kOBv4lvZjnNJU9iSRrgV8DDkvyLWA1va4cVfVf6V1G/L+nd4mO\nb9Abm56EFwCvAW5vhr4A3gE8HXqxVtWVSV6WZBPwE+B10xhnksOBfwSeDDyc5K3AijEn33njBP4Y\nOAR4f/O5sb2qVk5ZjL8F/F6S7cCP6X1jG7dB4pwGg8T528Abk+wAfsqUvp5VtTHJVcDtwMPAB6pq\nrElikDibbecAVze9njn5YzpJUqtFOdwkSRoPk4QkqZVJQpLUyiQhSWplkpAktTJJSJJaTeXvJKRp\nkeQpwLXN3cPpXf7lXuBZwJqqelNT7/HA3cADu6hbwMqq2jHG0KWh8HcS0oCSrAZ+VFV/sYvHTgfO\nqaq3zldXWkwcbpIWZuflITpJPtO3/Ux6V3x9TN2m/p8mubDv/kwmtLiTtBAmCWk4OkB3jsc/AZzb\nd/93gI+PMB5pKJyTkPZQsy7D/VX1YFudqro1ydOSHAE8Dfh+VY39asDSQpkkpD13JnDVAPU+Se9i\ndYdjL0KLhElC2nMvAd45QL1PAB8EnkLvMs3S1HNOQlqY6vu3kjwOeFZV3T1H3d6d3mWjDwK2jHtN\nEWl3eQqstAeSvAD43aq6cN7K0iJkkpAktXK4SZLUyiQhSWplkpAktTJJSJJamSQkSa1MEpKkViYJ\nSVKr/w9mcWsxHYxPdgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f7c9ca02690>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def calculate_titv(G):\n",
    "    expr = G.apply(\"alt_copies\", \"iif(allele_1,1,0)+ iif(allele_2,1,0)\")\n",
    "    expr = sdb.merge(expr, SNP['is_transition'])\n",
    "    expr = expr.apply(\"ti\", \"iif(is_transition, alt_copies, 0)\", \\\n",
    "                          \"tv\", \"iif(not is_transition, alt_copies, 0)\").\\\n",
    "            aggregate(\"sum(ti) as ti, sum(tv) as tv\", \"sample_id\").\\\n",
    "            eval()\n",
    "    return expr\n",
    "\n",
    "titv = calculate_titv(GENOTYPE).todataframe()\n",
    "\n",
    "plt.hist(titv['ti']/titv['tv'])\n",
    "plt.title(\"\")\n",
    "plt.xlabel(\"Ti/Tv\")\n",
    "plt.ylabel(\"# individuals\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All the data traversal is performed by SciDB in parallel; then the histogram buckets and counts are imported into R memory for plotting. Note that lazily-evaluated subexpressions can be stored, composed, and reused as R variables. This is often a lot nicer than formulating SQL!\n",
    "\n",
    "We can also calculate Ti/Tv for just one individual, by filtering the input matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample_id\n",
       "1234         2.275641\n",
       "dtype: float64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "titv1 = calculate_titv(\n",
    "    sdb.merge(\n",
    "        GENOTYPE, \n",
    "        sdb.afl.filter(SAMPLE['sample'], \"sample = 'HG03209'\")\n",
    "    )\n",
    "    ).todataframe()\n",
    "titv1['ti']/titv1['tv']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Principal component analysis\n",
    "\n",
    "Now let's find principal components of the genotype data and project the individual genomes onto them, revealing the underlying population structure. Begin by selecting common SNPs, because rare variants inherently don't contribute much to the overall variance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# TBD\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The redimension operation just removes some unneeded attributes and dimensions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# # Commented in original file \n",
    "# # downsampling for test/dev:\n",
    "# # SNP_COMMON <- scidbeval(bernoulli(SNP_COMMON, 0.1, 42))\n",
    "# # count(SNP_COMMON)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now construct a matrix `G` where `G[i,j]` is the number of copies of the alt allele found in sample `i`, common SNP `j`. The samples are observations, and the SNPs are variables. Reference: doi:10.1371/journal.pgen.0020190\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# TBD"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
